### PR TITLE
Feature/role instructor

### PR DESCRIPTION
**Pull request by VitaNuova**
Pull request for the approach to mapping features to endpoints and checking access rights to particular courses/exercises

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -67,6 +67,7 @@ local.properties
 *.ids
 *.orig
 classes/
+.idea/
```
Comment (krusche) [inline .gitignore]: please remove this change

File: .jhipster/Course.json [hunk 2]
```diff
@@ -20,11 +20,28 @@
...
+        {
+            "fieldName": "startDate",
+            "fieldType": "ZonedDateTime"
+        },
+        {
+            "fieldName": "endDate",
+            "fieldType": "ZonedDateTime"
+        },
+        {
+            "fieldName": "onlineCourse",
```
Comment (krusche) [inline .jhipster/Course.json]: Can you rename it to "isOnlineCourse"?

File: src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public boolean isAuthorizedForCourse(Course course) {
+        if(course == null) {
+            return true;
+        }
+        User user = userService.getUserWithGroupsAndAuthorities();
+        if(user.getGroups().contains(course.getStudentGroupName())
+            || user.getGroups().contains(course.getTeachingAssistantGroupName())
+            || user.getGroups().contains(course.getInstructorGroupName())
+            || user.getAuthorities().contains(adminAuthority)
+            || course.getTitle().equals("Archive")) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java]: I am not sure where you got this code from. This seems to be old code that is not used any more... please have a look into the master branch

File: src/main/java/de/tum/in/www1/exerciseapp/service/CourseService.java [hunk 4]
```diff
@@ -48,19 +48,7 @@ public Course save(Course course) {
...
-        taAuthority.setName("ROLE_TA");
-        Stream<Course> userCourses = result.stream().filter(
-            course -> user.getGroups().contains(course.getStudentGroupName())
-                || user.getGroups().contains(course.getTeachingAssistantGroupName())
-                || user.getAuthorities().contains(adminAuthority)
-        );
-        List<Course> userAuthorizedCourses = userCourses.collect(Collectors.toList());
-        return userAuthorizedCourses;
+        return courseRepository.findAll();
     }
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/service/CourseService.java]: why did you remove all the logic? I don't understand it

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java [hunk 5]
```diff
@@ -246,7 +246,7 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
      * @return the ResponseEntity with status 200 (OK)
      */
     @DeleteMapping(value = "/exercises/{id}/participations")
-    @PreAuthorize("hasAnyRole('ADMIN', 'TA')")
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: TA is missing.

File: src/main/webapp/i18n/de/course.json [hunk 6]
```diff
@@ -18,6 +18,10 @@
             "title": "Titel",
             "studentGroupName": "Studenten Gruppen Name",
             "teachingAssistantGroupName": "Tutor Gruppen Name",
+            "instructorGroupName": "Lehrer Gruppenname",
```
Comment (krusche) [inline src/main/webapp/i18n/de/course.json]: Instead of "Lehrer" please use "Dozent"

File: src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    private final UserService userService;
+    private Authority adminAuthority;
+
+    public AuthorizationCheckService(UserService userService) {
+        this.userService = userService;
+        adminAuthority = new Authority();
+        adminAuthority.setName("ROLE_ADMIN");
+    }
+
+    public boolean isAuthorizedForCourse(Course course) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java]: please add JavaDoc comments

File: src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java (hunk 3 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java]: at least the "Archive" option should not be used any more

Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java]: The "Archive" option is used at least in the findAll() method in the ExerciseService, that's why I believed it was relevant. The rest are from the findAll() method of CourseService, which you showed me yourself.

File: src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java (hunk 7 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/service/AuthorizationCheckService.java]: Ok

File: src/main/webapp/i18n/de/course.json (hunk 6 above)
Comment (VitaNuova) [reply src/main/webapp/i18n/de/course.json]: Ok

File: .gitignore (hunk 1 above)
Comment (VitaNuova) [reply .gitignore]: Ok

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 5 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: Reset feature is currently not in the list of features. I've added it now to the tables with the access right 'TA', hopefully that's correct.

File: src/main/java/de/tum/in/www1/exerciseapp/service/ExerciseService.java [hunk 8]
```diff
@@ -138,11 +138,6 @@ public void reset(Exercise exercise) {
     @Transactional
     public void delete(Long id) {
         log.debug("Request to delete Exercise : {}", id);
-        Exercise exercise = exerciseRepository.findOne(id);
-
-        if (Optional.ofNullable(exercise).isPresent()) {
-            reset(exercise);
-        }
         exerciseRepository.delete(id);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/service/ExerciseService.java]: Why did you remove this code? It is essential, and the delete will otherwise not work. Please add it back!

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java [hunk 9]
```diff
@@ -77,7 +73,7 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
     @PostMapping("/exercises")
-    @PreAuthorize("hasAnyRole('ADMIN', 'TA')")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     //TODO: test if it still works with abstract entity in body
     public ResponseEntity<Exercise> createExercise(@RequestBody Exercise exercise) throws URISyntaxException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: This should only be possible for TAs and instrcutors of this particular course (and admins)

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java [hunk 10]
```diff
@@ -121,7 +117,7 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
     @PutMapping("/exercises")
-    @PreAuthorize("hasAnyRole('ADMIN', 'TA')")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     //TODO: test if it still works with abstract entity in body
     public ResponseEntity<Exercise> updateExercise(@RequestBody Exercise exercise) throws URISyntaxException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: This should only be possible for TAs and instrcutors of this particular course (and admins)

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java [hunk 11]
```diff
@@ -148,7 +144,7 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
      * @return the ResponseEntity with status 200 (OK) and the list of exercises in body
      */
     @GetMapping("/exercises")
-    @PreAuthorize("hasAnyRole('ADMIN', 'TA')")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     public ResponseEntity<List<Exercise>> getAllExercises(@ApiParam Pageable pageable) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: This should only be possible for TAs and instrcutors of this particular course (and admins)

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java [hunk 12]
```diff
@@ -164,38 +160,25 @@ public ExerciseResource(ExerciseRepository exerciseRepository, ExerciseService e
...
+        if(user.getGroups().contains(course.getStudentGroupName())) {
             // user is student for this course
             result = withLtiOutcomeUrlExisting ? exerciseRepository.findByCourseIdWhereLtiOutcomeUrlExists(courseId, principal) : exerciseRepository.findByCourseId(courseId);
             // filter out exercises that are not released (or explicitly made visible to students) yet
             result = result.stream().filter(Exercise::getIsVisibleToStudents).collect(Collectors.toList());
         } else {
-            // user has no permissions for this course
-            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+            result = exerciseRepository.findByCourseId(courseId);
         }
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: are you 100% sure that the logic stays the same with the new service? This is extremely important, because this call is invoked on the start page for all ArTEMiS users and we spent a lot of time to get it right

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java [hunk 13]
```diff
@@ -74,7 +68,7 @@ public ParticipationResource(ParticipationService participationService, Particip
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
     @PostMapping("/participations")
-    @PreAuthorize("hasAnyRole('TA', 'ADMIN')")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     public ResponseEntity<Participation> createParticipation(@RequestBody Participation participation) throws URISyntaxException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java]: This should only be possible for TAs and instrcutors of this particular course (and admins)

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java [hunk 14]
```diff
@@ -144,7 +144,7 @@ public ParticipationResource(ParticipationService participationService, Particip
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
     @PutMapping("/participations")
-    @PreAuthorize("hasAnyRole('TA', 'ADMIN')")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     public ResponseEntity<Participation> updateParticipation(@RequestBody Participation participation) throws URISyntaxException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java]: This should only be possible for TAs and instrcutors of the corresponding exercise (and admins)

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java [hunk 15]
```diff
@@ -51,7 +51,7 @@ public QuizExerciseResource(QuizExerciseRepository quizExerciseRepository, Parti
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java]: please apply the same logic here as for programming exercises

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java [hunk 16]
```diff
@@ -300,7 +320,7 @@ public ResultResource(ResultRepository resultRepository, Optional<LtiService> lt
      * @return the ResponseEntity with status 200 (OK)
      */
     @DeleteMapping("/results/{id}")
-    @PreAuthorize("hasAnyRole('TA', 'ADMIN')")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     public ResponseEntity<Void> deleteResult(@PathVariable Long id) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: This should only be possible for TAs and instrcutors of the corresponding exercise (and admins)

File: src/main/java/de/tum/in/www1/exerciseapp/service/ExerciseService.java (hunk 8 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/service/ExerciseService.java]: I moved this code to the ExerciseResource, please check line 227.  I anyway need to retrieve Exercise from ExerciseRepository in order to check authorization and, in order not to call ExerciseRepository.findOne(id) twice, I've moved this code to the resource.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 9 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: Create exercise feature is currently using POST /api/programming-exercises, where I have inserted the check. I wasn't able to find the usage of this method anywhere and, therefore, there is no check.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 10 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: Same here. Edit exercise feature is using PUT /api/programming-exercises. I wasn't able to identify the use of this method.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 11 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: I wasn't able to identify where this method is used. According to my observations, GET /course/:courseId/exercises is used to retrieve exercises for a particular course, and this is done for all courses. Do you know the usage of this method?

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 12 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: As far as I can judge.. If you look into the new service, it does all the same checks except user.getAuthorities().contains(taAuthority), which I find excessive because we anyway do user.getGroups().contains(course.getTeachingAssistantGroupName()). The new service also adds the check for the instructor role. I have checked for myself (once being admin and once not), and it resolves courses fine, but I obviously can't check for anyone else.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java (hunk 13 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java]: Where is this used? I suspect it was generated with JHipster and never used. Participations are created with POST /api/courses/:id/exercises/:id/participations.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java (hunk 14 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java]: Same here. I believe this one was also never touched after having been generated.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java (hunk 15 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java]: Didn't you say that quiz and modeling exercises should be done by the respective authors? I asked you when I wasn't able to map quiz endpoints, and I believed you said something along the lines that the respective authors should do that. Should I then just blindly add checks to all endpoints whether they are used or not?

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java (hunk 16 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: Same. I believe we have no feature to delete results from user interface.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 9 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: Done.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 10 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: Done.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java (hunk 11 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ExerciseResource.java]: Done.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java (hunk 13 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java]: Done.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java (hunk 14 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ParticipationResource.java]: Done,

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java (hunk 16 above)
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: Done.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/CourseResource.java [hunk 17]
```diff
@@ -34,9 +38,11 @@
     private static final String ENTITY_NAME = "course";
 
     private final CourseService courseService;
+    private final AuthorizationCheckService authCheckService;
 
-    public CourseResource(CourseService courseService) {
+    public CourseResource(CourseService courseService, AuthorizationCheckService authCheckService) {
```
Comment (sockeqwe) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/CourseResource.java]: @VitaNuova just FYI:
Project is not compiling anymore because unit tests are not compiling anymore. Unit test still use the old constructor, see here: [CourseResourceIntTest](https://github.com/ls1intum/ArTEMiS/blob/develop/src/test/java/de/tum/in/www1/exerciseapp/web/rest/CourseResourceIntTest.java#L93)

I have fixed that on my branch (hd/umleditor) but I'm wondering why the code is compiling on your machine 🤔

Comment (krusche) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/CourseResource.java]: Thanks for the hint @sockeqwe I fixed it in the develop branch as well with my latest commit.
